### PR TITLE
UAF-3787 Automatic Prorate Journal Entries (PRJE) Step 2 (Reporting, …

### PR DESCRIPTION
**Pull request by mmoen**
…Batch Jobs, Audit Item Business Object and Data Access Objects)

Modified KFSConstants.java to add a constant for use in other files.
Modified PRJEPropertyConstants.java to add constants for use in other files.
Added new file PRJETransactionStep.java. Batch step to generate PRJE transactions based on the PRJE type table.
Added new file PRJEAuditItem.java. A helper class that's only used in PRJE Reporting.
Added new file PRJEDao.java. An interface that declares methods needed for PRJE to interact with the database
Added new file PRJETransferRecord.java. Encapsulates Withdrawal and Deposit data for PRJE Transactions.  Used as a common representation of transfer data between the PRJE Service and Reporting system.
Added new file PRJEDaoOjb.java that implements the PRJEDao interface.
Added new file PRJEReportsService.java. An interface that declares methods needed to write reports.
Added new file PRJEServiceBase.java and PRJEService.java that define the methods that does the actual work of PRJETransactionStep.
Added new file PRJEReportsServiceImpl.java that implements the PRJEReportsService interface.
Added new file PRJEServiceImpl.java that implements the PRJEService interface.
Added new file PRJEAuditItem.xml to define the beans for PRJEAuditItem.
Added new file dwr-prje.xml to dynamically generate JavaScript code based on the Java class.
Modified spring-prje.xml to add beans for the new classes.

### CODE REVIEW

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package edu.arizona.kfs.module.prje.batch;
+
+import java.util.Date;
+
+import org.kuali.kfs.sys.batch.AbstractStep;
+import org.kuali.kfs.sys.batch.Step;
+
+import edu.arizona.kfs.module.prje.service.PRJEService;
+
+public class PRJETransactionStep extends AbstractStep implements Step {
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java]: No need to implement Step as AbstractStep already does that.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package edu.arizona.kfs.module.prje.batch;
+
+import java.util.Date;
+
+import org.kuali.kfs.sys.batch.AbstractStep;
+import org.kuali.kfs.sys.batch.Step;
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java]: Remove import, please. No need to implement Step as AbstractStep already does that.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+import edu.arizona.kfs.module.prje.service.PRJEService;
+
+public class PRJETransactionStep extends AbstractStep implements Step {
+    private PRJEService prjeService;
+    
+    public boolean execute(String jobName, Date jobRunDate) throws InterruptedException {
+        return getPrjeService().process();
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java]: Getter method not needed. Should be: return prjeService.process();

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    public PRJEService getPrjeService() {
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java]: Getter method not needed. Please remove.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/businessobject/PRJEAuditItem.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
+package edu.arizona.kfs.module.prje.businessobject;
+
+import java.util.LinkedHashMap;
+
+import org.kuali.kfs.gl.businessobject.OriginEntryFull;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+import org.kuali.rice.krad.bo.PersistableBusinessObjectBase;
+
+import edu.arizona.kfs.module.prje.PRJEPropertyConstants;
+import edu.arizona.kfs.module.prje.dataaccess.PRJETransferRecord;
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/businessobject/PRJEAuditItem.java]: PRJETransferRecord looks like a BO to me. Why in dataaccess package?

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/businessobject/PRJEAuditItem.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+import edu.arizona.kfs.module.prje.PRJEConstants.ProrateCreditType;
+
+/**
+ * This class is only used in PRJE Reporting
+ */
+public class PRJEAuditItem extends PersistableBusinessObjectBase {
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/businessobject/PRJEAuditItem.java]: Clearly not a PersistableBusinessObject as there is no OJB definition. Did TransientBusinessObjectBase not work? Or no extension at all?

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/dataaccess/PRJEDao.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package edu.arizona.kfs.module.prje.dataaccess;
+
+import java.util.Collection;
+
+import edu.arizona.kfs.module.prje.businessobject.PRJESet;
+import edu.arizona.kfs.module.prje.businessobject.PRJEType;
+
+public interface PRJEDao {
+    Collection<PRJEType> getPRJETypes(PRJESet set, Boolean active);
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/dataaccess/PRJEDao.java]: Why no access level modifiers for these methods? Would like to see public access level modifier used here.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/dataaccess/PRJETransferRecord.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+import edu.arizona.kfs.module.prje.businessobject.PRJEAccountLine;
+import edu.arizona.kfs.module.prje.businessobject.PRJEBaseAccount;
+
+/**
+ * Encapsulates Withdrawal and Deposit data for PRJE Transactions.  Used as
+ * a common representation of transfer data between the PRJE Service and
+ * Reporting system.
+ */
+public class PRJETransferRecord {
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/dataaccess/PRJETransferRecord.java]: This class looks like a transient BO to me. Why in dataaccess package?

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEReportsService.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.List;
+
+import edu.arizona.kfs.module.prje.dataaccess.PRJETransferRecord;
+
+public interface PRJEReportsService {
+    /**
+     * Will generate all of the Reports made available by this module
+     * @param transferRecords
+     */
+    void writeReports(List<PRJETransferRecord> transferRecords);
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEReportsService.java]: Why no access level modifiers for these methods? Would like to see public access level modifier used here.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEService.java [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package edu.arizona.kfs.module.prje.service;
+
+public interface PRJEService {
+    boolean process();
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEService.java]: Why no access level modifier for this method? Would like to see public access level modifier used here.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
+package edu.arizona.kfs.module.prje.service;
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java]: If other service impl classes such as PRJEReportsServiceImpl.java extend this class, this class should also be an "impl" class (edu.arizona.kfs.module.prje.service.impl.PRJEServiceBaseImpl.java).

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+import org.kuali.kfs.coa.service.ObjectCodeService;
+import org.kuali.kfs.gl.GeneralLedgerConstants;
+import org.kuali.kfs.sys.context.SpringContext;
+import org.kuali.kfs.sys.service.UniversityDateService;
+import org.kuali.rice.core.api.datetime.DateTimeService;
+import org.kuali.rice.coreservice.framework.parameter.ParameterService;
+import org.kuali.rice.krad.service.BusinessObjectService;
+import org.kuali.rice.krad.service.KRADServiceLocator;
+
+import edu.arizona.kfs.module.prje.PRJEConstants;
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java]: import not needed.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,857 @@
...
+            // or debit the departmental account for revenue object 
+            // codes.
+
+            // Always post the prorated amounts to the account and object
+            // code (no sub-account and sub-object codes) since they are
+            // not guaranteed to exist.
+
+            // Create the origin entry
+            OriginEntryFull creditEntry = new OriginEntryFull(PRJEConstants.ENTRY_DOCTYPE, PRJEConstants.ENTRY_ORIGIN);
+            creditEntry.setFinancialBalanceTypeCode(KFSConstants.BALANCE_TYPE_ACTUAL);
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: KFSConstants.BALANCE_TYPE_ACTUAL is deprecated. You should get these values from the system options service. See  edu.arizona.kfs.gl.batch.service.impl.BudgetAdjustmentCashTransferServiceImpl.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,857 @@
...
+                chartCode = baseAccount.getBaseChart();
+            }
+            
+            LOG.info("--- Debit " + amount + " from Base Account " + account);
+            
+            PRJEType type = baseAccount.getType();
+            
+            // Create the origin entry
+            OriginEntryFull debitEntry = new OriginEntryFull(PRJEConstants.ENTRY_DOCTYPE, PRJEConstants.ENTRY_ORIGIN);
+            debitEntry.setFinancialBalanceTypeCode(KFSConstants.BALANCE_TYPE_ACTUAL);
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: KFSConstants.BALANCE_TYPE_ACTUAL is deprecated. You should get these values from the system options service. See edu.arizona.kfs.gl.batch.service.impl.BudgetAdjustmentCashTransferServiceImpl.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,857 @@
...
+            
+            return result;
+        }
+        
+        private KualiDecimal normalizeBalance(Balance balance, KualiDecimal amount) {
+            return amount;
+        }
+        
+        private boolean balanceMatchesBaseObjects(PRJEType type, Balance balance) {
+            if ( !KFSConstants.BALANCE_TYPE_ACTUAL.equals(balance.getBalanceTypeCode()) ) {
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: KFSConstants.BALANCE_TYPE_ACTUAL is deprecated. You should get these values from the system options service. See edu.arizona.kfs.gl.batch.service.impl.BudgetAdjustmentCashTransferServiceImpl.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,857 @@
...
+        }
+        
+        private KualiDecimal calculateBaseAccountAmount(PRJEBaseAccount baseAccount) {
+            KualiDecimal curBalance = ZERO;
+            LOG.debug("-- Processing Base Account "+baseAccount.getBaseAccount());
+            
+            // Retrieve the balances for this baseAccount
+            BusinessObjectService bos = getBusinessObjectService();
+            Map<String, String> args = new HashMap<String, String>();
+            args.put("universityFiscalYear", Integer.toString(fiscalYear));
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: Should use KFSPropertyConstants.UNIVERSITY_FISCAL_YEAR, KFSPropertyConstants.CHART_OF_ACCOUNTS_CODE and KFSPropertyConstants.ACCOUNT_NUMBER here.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,857 @@
...
+            args.put("chartOfAccountsCode", baseAccount.getBaseChart());
+            args.put("accountNumber", baseAccount.getBaseAccount());
+            
+            if ( baseAccount.getBaseSubAccount() != null ) {
+                args.put("subAccountNumber", baseAccount.getBaseSubAccount());
```
Comment (ghost) [inline kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: Should use KFSPropertyConstants.SUB_ACCOUNT_NUMBER here.

File: kfs-prje/src/main/resources/edu/arizona/kfs/module/prje/spring-prje.xml [hunk 18]
```diff
@@ -27,10 +27,186 @@
...
+			</list>
+		</property>
+	</bean>
+
+	<bean id="prjeTransactionStep" parent="step"
+	      class="edu.arizona.kfs.module.prje.batch.PRJETransactionStep">
+		<property name="prjeService" ref="prjeService" />
+	</bean>
+
+	<bean id="prjeJobTrigger" parent="cronTrigger">
```
Comment (ghost) [inline kfs-prje/src/main/resources/edu/arizona/kfs/module/prje/spring-prje.xml]: I don't believe this trigger is needed as the prjeJob (as all our jobs are) is unscheduled.

**Review by ghost — COMMENTED** [review]
Initial review complete. Please see my comments.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java (hunk 2 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java (hunk 1 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java (hunk 3 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java (hunk 4 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/batch/PRJETransactionStep.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/businessobject/PRJEAuditItem.java (hunk 5 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/businessobject/PRJEAuditItem.java]: I didn't notice it was in the wrong package when I took over the mod. Moving to businessobject package

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/businessobject/PRJEAuditItem.java (hunk 6 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/businessobject/PRJEAuditItem.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/dataaccess/PRJEDao.java (hunk 7 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/dataaccess/PRJEDao.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/dataaccess/PRJETransferRecord.java (hunk 8 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/dataaccess/PRJETransferRecord.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEReportsService.java (hunk 9 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEReportsService.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEService.java (hunk 10 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEService.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java (hunk 11 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java (hunk 12 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java (hunk 14 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java (hunk 13 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java (hunk 16 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java (hunk 17 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: fixed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java (hunk 15 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/impl/PRJEServiceImpl.java]: fixed

File: kfs-prje/src/main/resources/edu/arizona/kfs/module/prje/spring-prje.xml (hunk 18 above)
Comment (mmoen) [reply kfs-prje/src/main/resources/edu/arizona/kfs/module/prje/spring-prje.xml]: removed

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java (hunk 11 above)
Comment (ghost) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java]: Kudos for renaming this, but it should be in "impl" package:
(edu.arizona.kfs.module.prje.service.impl.PRJEServiceBaseImpl.java).

**Review by ghost — COMMENTED** [review]
Second review complete. Please see my comment.

File: kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java (hunk 11 above)
Comment (mmoen) [reply kfs-prje/src/main/java/edu/arizona/kfs/module/prje/service/PRJEServiceBase.java]: fixed

**Comment by ghost** [comment]
Looks good! Waiting for final prototype env testing to be completed before merging.

**Comment by ghost** [comment]
Merging.

**Pull request merged**